src/tests/ftest/cart/utest/utest_portnumber.c

```diff
@@ -0,0 +1,324 @@
+			if (rc != 0) {
+				rc = CHILD2_CONTEXT_DESTROY_ERR;
+				rc = crt_finalize();
+				goto child2_error;
+			}
+		}
+
+		/* signal child 1 to finish up */
+		sem_post(child1_sem);
+		rc = crt_finalize();
+		if (rc != 0) {
+			rc = CHILD2_FINALIZE_ERR;
+			goto child2_error;
+		}
+		exit(child_result);
+child2_error:
+		sem_post(child1_sem);
+		exit(rc);
+	}
+
+	/* Wait for first child and get results */
+	waitpid(pid1, &status, 0);
+	if (WIFEXITED(status)) {
+		result1 = WEXITSTATUS(status);
+	}
+
+	/* Wait for second child and get results */
+	waitpid(pid2, &status, 0);
+	if (WIFEXITED(status)) {
+		result2 = WEXITSTATUS(status);
+	}
+
+	/* Test results.  first child should should succeed. */
+	assert_true(result1 == 0);
+	assert_true(result2 != 0);
+	assert_false(result2 == CHILD2_INIT_ERR);
+	assert_false(result2 == CHILD2_CONTEXT_DESTROY_ERR);
+	assert_false(result2 == CHILD2_FINALIZE_ERR);
+	assert_false(result2 == CHILD2_TIMEOUT_ERR);
+}
+
+static void
+test_port_tcp(void **state)
+{
+	setenv("OFI_INTERFACE", "lo", 1);
+	setenv("CRT_PHY_ADDR_STR", "ofi+tcp;ofi_rxm", 1);
+	run_test_fork(state);
+}
+
+#ifndef MY_TESTS_NOT_INCLUDED
+static void
+test_port_sockets(void **state)
+{
+	setenv("OFI_INTERFACE", "eth0", 1);
+	setenv("CRT_PHY_ADDR_STR", "ofi+sockets", 1);
+	run_test_fork(state);
+};
+
+static void
+test_port_verb(void **state)
+{
+	setenv("OFI_INTERFACE", "eth0", 1);
+	setenv("OFI_DOMAIN", "Must define here", 1);
+	setenv("CRT_PHY_ADDR_STR", "ofi+verbs;ofi_rxm", 1);
+	run_test_fork(state);
+};
+#endif
+
+/*******************/
+static int
+init_tests(void **state)
+{
+	int size = sizeof(sem_t);
+	int flag = IPC_CREAT | 0666;
+	int pshare = 1;
+	int init_value = 1;
+	int rc;
+
+	/* create shared memory regions for semaphores */
+	shmid_c1 = shmget(IPC_PRIVATE, size, flag);
+	assert_true(shmid_c1  != -1);
+	shmid_c2 = shmget(IPC_PRIVATE, size, flag);
+	assert_true(shmid_c2  != -1);
+	if ((shmid_c1 == -1) || (shmid_c2 == -1)) {
+		printf(" SHMID not created\n");
+		goto cleanup;
+	}
+
+	/* Attach share memory regions */
+	child1_sem = (sem_t *)shmat(shmid_c1, NULL, 0);
+	child2_sem = (sem_t *)shmat(shmid_c2, NULL, 0);
+
+	assert_true(child1_sem  != (sem_t *)-1);
+	assert_true(child2_sem  != (sem_t *)-1);
+
+	/* Initialize semaphores */
+	rc = sem_init(child1_sem, pshare, init_value);
+	assert_true(rc == 0);
+
+	rc = sem_init(child2_sem, pshare, init_value);
+	assert_true(rc == 0);
+
+	return 0;
+
+cleanup:
+	printf(" Error creating semaphores\n");
+	return -1;
+}
+
+/*******************/
+static int
+fini_tests(void **state)
+{
+	struct shmid_ds smds;
+	int rc;
+
+	/* detach shared memory region */
+	if (child1_sem != NULL)
+		shmdt(child1_sem);
+	if (child2_sem != NULL)
+		shmdt(child2_sem);
+
+	/* eliminate the shared memory regions */
+	rc = shmctl(shmid_c1, IPC_STAT, &smds);
+	assert_true(rc == 0);
+	rc = shmctl(shmid_c1, IPC_RMID, &smds);
+	assert_true(rc == 0);
+
+	rc = shmctl(shmid_c2, IPC_STAT, &smds);
```

Comment:
I don't think cmocka assertions should be used in init and fini routines.  It depends on how setjmp and longjmp are used.  Even if they are safe, this can cause a resource leak if it explicitly aborts.   Instead, we should try to clean up as much as possible even if one of them fails for some reason.

Reply:
You might try a test program where you assert in fini and see what happens.